cpp/ql/lib/semmle/code/cpp/ir/implementation/aliased_ssa/internal/SSAConstruction.qll

```diff
@@ -371,6 +498,35 @@ private module Cached {
       )
   }
 
+  private Instruction getInstructionSuccessorAfterUninitializedGroup(
+    Instruction instruction, EdgeKind kind
+  ) {
+    exists(IRFunction func, Instruction firstBeforeUninitializedGroup |
+      isLastInstructionForUninitializedGroups(instruction, func) and
+      isFirstInstructionBeforeUninitializedGroup(firstBeforeUninitializedGroup, func) and
+      result = getInstructionSuccessorAfterUninitializedGroup0(firstBeforeUninitializedGroup, kind)
+    )
+  }
+
+  /**
+   * This adds Chi nodes to the instruction successor relation; if an instruction has a Chi node,
+   * that node is its successor in the new successor relation, and the Chi node's successors are
+   * the new instructions generated from the successors of the old instruction.
+   *
+   * Furthermore, the entry block is augmented with `UninitializedGroup` instructions.
+   */
+  cached
+  Instruction getInstructionSuccessor(Instruction instruction, EdgeKind kind) {
+    result = firstInstructionToUninitializedGroup(instruction, kind)
+    or
+    result = getNextUninitializedGroupInstruction(instruction, kind)
+    or
+    result = getInstructionSuccessorAfterUninitializedGroup(instruction, kind)
+    or
+    not isFirstInstructionBeforeUninitializedGroup(instruction, _) and
+    result = getInstructionSuccessorAfterUninitializedGroup0(instruction, kind)
+  }
```

Comment:
The instruction ordering code has accumulated a very high level of complexity for what it is, and I feel this is error prone (and difficult to maintain).  I think we should discuss ways we can improve this in future, in particular the simple operation of appending one variable length block of instructions to another creates quite a lot of code / complexity at the moment.

Reply:
Yeah, it would be good if we could do something prettier here in the future. It's _doable_ for now since we're only injecting new `UninitializedGroup` instructions and `ChiInstruction`s (and because we know exactly the structure of these), but I agree that this could get unmanageable

Reply:
I would actually say it's already unmanageable, at least to my tolerance.  I'm not suggesting we fix it here, I'll create an issue for discussion...

Reply:
I'll get started on figuring out a more sensible way to structure this as a follow-up PR

Reply:
(I didn't actually create the issue, but we discussed stuff on another channel)